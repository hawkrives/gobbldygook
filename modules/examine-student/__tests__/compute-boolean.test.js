import computeChunk, { computeBoolean } from '../compute-chunk'

describe('computeBoolean', () => {
    it('computes the boolean result of and-clauses', () => {
        const clause = {
            $type: 'boolean',
            $booleanType: 'and',
            $and: [
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 121 },
                },
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 125 },
                },
            ],
        }
        const requirement = { result: clause }
        const courses = [
            { department: ['CSCI'], number: 121 },
            { department: ['CSCI'], number: 125 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('computes the boolean result of or-clauses', () => {
        const clause = {
            $type: 'boolean',
            $booleanType: 'or',
            $or: [
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 121 },
                },
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 125 },
                },
            ],
        }
        const requirement = { result: clause }
        const courses = [
            { department: ['CSCI'], number: 121 },
            { department: ['CSCI'], number: 125 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('computes an or-clause even if the first item is false', () => {
        const clause = {
            $type: 'boolean',
            $booleanType: 'or',
            $or: [
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 121 },
                },
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 125 },
                },
            ],
        }
        const requirement = { result: clause }
        const courses = [
            { department: ['CSCI'], number: 151 },
            { department: ['CSCI'], number: 125 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several other boolean expressions', () => {
        const clause = {
            $type: 'boolean',
            $booleanType: 'and',
            $and: [
                {
                    $type: 'boolean',
                    $booleanType: 'or',
                    $or: [
                        {
                            $type: 'course',
                            $course: { department: ['CSCI'], number: 121 },
                        },
                        {
                            $type: 'course',
                            $course: { department: ['CSCI'], number: 125 },
                        },
                    ],
                },
                {
                    $type: 'boolean',
                    $booleanType: 'or',
                    $or: [
                        {
                            $type: 'course',
                            $course: { department: ['CSCI'], number: 130 },
                        },
                        {
                            $type: 'course',
                            $course: { department: ['CSCI'], number: 131 },
                        },
                    ],
                },
            ],
        }
        const requirement = { result: clause }

        const courses = [
            { department: ['CSCI'], number: 130 },
            { department: ['CSCI'], number: 125 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several course expressions', () => {
        const clause = {
            $or: [
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 121 },
                },
                {
                    $type: 'course',
                    $course: { department: ['CSCI'], number: 125 },
                },
            ],
            $type: 'boolean',
            $booleanType: 'or',
        }
        const requirement = { result: clause }

        const courses = [
            { department: ['CSCI'], number: 130 },
            { department: ['CSCI'], number: 125 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several modifier expressions', () => {
        const clause = {
            $and: [
                {
                    $children: '$all',
                    $count: { $operator: '$gte', $num: 3 },
                    $from: 'children',
                    $type: 'modifier',
                    $what: 'course',
                },
                {
                    $children: [
                        {
                            $requirement: 'A',
                            $type: 'reference',
                        },
                        {
                            $requirement: 'C',
                            $type: 'reference',
                        },
                    ],
                    $count: { $operator: '$gte', $num: 2 },
                    $from: 'children',
                    $type: 'modifier',
                    $what: 'credit',
                },
            ],
            $type: 'boolean',
            $booleanType: 'and',
        }
        const requirement = {
            A: {
                $type: 'requirement',
                result: {
                    $type: 'course',
                    $course: { department: ['ART'], number: 120 },
                },
            },
            C: {
                $type: 'requirement',
                result: {
                    $count: { $operator: '$gte', $num: 2 },
                    $of: [
                        {
                            $type: 'course',
                            $course: { department: ['ART'], number: 103 },
                        },
                        {
                            $type: 'course',
                            $course: { department: ['ART'], number: 104 },
                        },
                        {
                            $type: 'course',
                            $course: { department: ['ART'], number: 105 },
                        },
                    ],
                    $type: 'of',
                },
            },
            result: clause,
        }

        const courses = [
            { department: ['ART'], number: 120, credits: 1.0 },
            { department: ['ART'], number: 104, credits: 1.0 },
            { department: ['ART'], number: 105, credits: 1.0 },
        ]
        const dirty = new Set()

        requirement.A.computed = computeChunk({
            expr: requirement.A.result,
            ctx: requirement,
            courses,
            dirty,
        })
        requirement.C.computed = computeChunk({
            expr: requirement.C.result,
            ctx: requirement,
            courses,
            dirty,
        })

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty,
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several occurrence expressions', () => {
        const clause = {
            $or: [
                {
                    $count: { $operator: '$gte', $num: 1 },
                    $course: { department: ['THEAT'], number: 222 },
                    $type: 'occurrence',
                },
                {
                    $count: { $operator: '$gte', $num: 3 },
                    $course: { department: ['THEAT'], number: 266 },
                    $type: 'occurrence',
                },
            ],
            $type: 'boolean',
            $booleanType: 'or',
        }

        const requirement = { result: clause }

        const courses = [
            { department: ['THEAT'], number: 266, year: 2014, semester: 1 },
            { department: ['THEAT'], number: 266, year: 2014, semester: 3 },
            { department: ['THEAT'], number: 266, year: 2015, semester: 1 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several of-expressions', () => {
        const clause = {
            $and: [
                {
                    $count: { $operator: '$gte', $num: 1 },
                    $of: [
                        {
                            $course: { department: ['CSCI'], number: 121 },
                            $type: 'course',
                        },
                        {
                            $course: { department: ['CSCI'], number: 125 },
                            $type: 'course',
                        },
                    ],
                    $type: 'of',
                },
                {
                    $count: { $operator: '$gte', $num: 1 },
                    $of: [
                        {
                            $course: { department: ['ART'], number: 102 },
                            $type: 'course',
                        },
                        {
                            $course: { department: ['ART'], number: 103 },
                            $type: 'course',
                        },
                    ],
                    $type: 'of',
                },
            ],
            $type: 'boolean',
            $booleanType: 'and',
        }

        const requirement = { result: clause }

        const courses = [
            { department: ['CSCI'], number: 125 },
            { department: ['ART'], number: 102 },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several requirement references', () => {
        const clause = {
            $and: [
                { $requirement: 'A', $type: 'reference' },
                { $requirement: 'C', $type: 'reference' },
            ],
            $type: 'boolean',
            $booleanType: 'and',
        }
        const requirement = {
            A: {
                $type: 'requirement',
                result: {
                    $type: 'course',
                    $course: { department: ['ART'], number: 120 },
                },
            },
            C: {
                $type: 'requirement',
                result: {
                    $count: { $operator: '$gte', $num: 2 },
                    $of: [
                        {
                            $type: 'course',
                            $course: { department: ['ART'], number: 103 },
                        },
                        {
                            $type: 'course',
                            $course: { department: ['ART'], number: 104 },
                        },
                        {
                            $type: 'course',
                            $course: { department: ['ART'], number: 105 },
                        },
                    ],
                    $type: 'of',
                },
            },
            result: clause,
        }

        const courses = [
            { department: ['ART'], number: 120 },
            { department: ['ART'], number: 104 },
            { department: ['ART'], number: 105 },
        ]
        const dirty = new Set()

        requirement.A.computed = computeChunk({
            expr: requirement.A.result,
            ctx: requirement,
            courses,
            dirty,
        })
        requirement.C.computed = computeChunk({
            expr: requirement.C.result,
            ctx: requirement,
            courses,
            dirty,
        })

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty,
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('can compute the result of several where-expressions', () => {
        const clause = {
            $and: [
                {
                    $count: { $operator: '$gte', $num: 1 },
                    $type: 'where',
                    $where: {
                        $key: 'gereqs',
                        $operator: '$eq',
                        $type: 'qualification',
                        $value: 'WRI',
                    },
                    $distinct: false,
                },
                {
                    $count: { $operator: '$gte', $num: 1 },
                    $type: 'where',
                    $where: {
                        $key: 'gereqs',
                        $operator: '$eq',
                        $type: 'qualification',
                        $value: 'BTS-T',
                    },
                    $distinct: false,
                },
            ],
            $type: 'boolean',
            $booleanType: 'and',
        }

        const requirement = { result: clause }

        const courses = [
            { department: ['CSCI'], number: 125, gereqs: ['WRI'] },
            { department: ['ART'], number: 102, gereqs: ['BTS-T'] },
        ]

        const { computedResult, matches } = computeBoolean({
            expr: clause,
            ctx: requirement,
            courses,
            dirty: new Set(),
            isNeeded: true,
        })
        expect(clause).toMatchSnapshot()
        expect(computedResult).toBe(true)
        expect(matches).toMatchSnapshot()
    })

    it('throws when neither $and nor $or were present', () => {
        expect(() =>
            computeBoolean({
                expr: { $neither: [] },
                isNeeded: true,
            })).toThrowError(TypeError)
    })
})
