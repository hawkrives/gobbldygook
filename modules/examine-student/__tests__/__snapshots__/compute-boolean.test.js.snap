// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeBoolean can compute the result of several course expressions 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
      "_checked": true,
      "_result": false,
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
      "_checked": true,
      "_request": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "_result": true,
      "_taken": true,
    },
  ],
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several course expressions 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 125,
  },
]
`;

exports[`computeBoolean can compute the result of several modifier expressions 1`] = `
Object {
  "$and": Array [
    Object {
      "$children": "$all",
      "$count": Object {
        "$num": 3,
        "$operator": "$gte",
      },
      "$from": "children",
      "$type": "modifier",
      "$what": "course",
      "_checked": true,
      "_counted": 3,
      "_matches": Array [
        Object {
          "_extraKeys": Array [
            "credits",
          ],
          "credits": 1,
          "department": Array [
            "ART",
          ],
          "number": 120,
        },
        Object {
          "_extraKeys": Array [
            "credits",
          ],
          "credits": 1,
          "department": Array [
            "ART",
          ],
          "number": 104,
        },
        Object {
          "_extraKeys": Array [
            "credits",
          ],
          "credits": 1,
          "department": Array [
            "ART",
          ],
          "number": 105,
        },
      ],
      "_result": true,
    },
    Object {
      "$children": Array [
        Object {
          "$requirement": "A",
          "$type": "reference",
        },
        Object {
          "$requirement": "C",
          "$type": "reference",
        },
      ],
      "$count": Object {
        "$num": 2,
        "$operator": "$gte",
      },
      "$from": "children",
      "$type": "modifier",
      "$what": "credit",
      "_checked": true,
      "_counted": 3,
      "_matches": Array [
        Object {
          "_extraKeys": Array [
            "credits",
          ],
          "credits": 1,
          "department": Array [
            "ART",
          ],
          "number": 120,
        },
        Object {
          "_extraKeys": Array [
            "credits",
          ],
          "credits": 1,
          "department": Array [
            "ART",
          ],
          "number": 104,
        },
        Object {
          "_extraKeys": Array [
            "credits",
          ],
          "credits": 1,
          "department": Array [
            "ART",
          ],
          "number": 105,
        },
      ],
      "_result": true,
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several modifier expressions 2`] = `
Array [
  Object {
    "_extraKeys": Array [
      "credits",
    ],
    "credits": 1,
    "department": Array [
      "ART",
    ],
    "number": 120,
  },
  Object {
    "_extraKeys": Array [
      "credits",
    ],
    "credits": 1,
    "department": Array [
      "ART",
    ],
    "number": 104,
  },
  Object {
    "_extraKeys": Array [
      "credits",
    ],
    "credits": 1,
    "department": Array [
      "ART",
    ],
    "number": 105,
  },
]
`;

exports[`computeBoolean can compute the result of several occurrence expressions 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$course": Object {
        "department": Array [
          "THEAT",
        ],
        "number": 222,
      },
      "$type": "occurrence",
      "_checked": true,
      "_counted": 0,
      "_matches": Array [],
      "_result": false,
    },
    Object {
      "$count": Object {
        "$num": 3,
        "$operator": "$gte",
      },
      "$course": Object {
        "department": Array [
          "THEAT",
        ],
        "number": 266,
      },
      "$type": "occurrence",
      "_checked": true,
      "_counted": 3,
      "_matches": Array [
        Object {
          "department": Array [
            "THEAT",
          ],
          "number": 266,
          "semester": 1,
          "year": 2014,
        },
        Object {
          "department": Array [
            "THEAT",
          ],
          "number": 266,
          "semester": 3,
          "year": 2014,
        },
        Object {
          "department": Array [
            "THEAT",
          ],
          "number": 266,
          "semester": 1,
          "year": 2015,
        },
      ],
      "_result": true,
    },
  ],
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several occurrence expressions 2`] = `
Array [
  Object {
    "department": Array [
      "THEAT",
    ],
    "number": 266,
    "semester": 1,
    "year": 2014,
  },
  Object {
    "department": Array [
      "THEAT",
    ],
    "number": 266,
    "semester": 3,
    "year": 2014,
  },
  Object {
    "department": Array [
      "THEAT",
    ],
    "number": 266,
    "semester": 1,
    "year": 2015,
  },
]
`;

exports[`computeBoolean can compute the result of several of-expressions 1`] = `
Object {
  "$and": Array [
    Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$of": Array [
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 121,
          },
          "$type": "course",
          "_checked": true,
          "_result": false,
        },
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 125,
          },
          "$type": "course",
          "_checked": true,
          "_request": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 125,
          },
          "_result": true,
          "_taken": true,
        },
      ],
      "$type": "of",
      "_checked": true,
      "_counted": 1,
      "_matches": Array [
        Object {
          "department": Array [
            "CSCI",
          ],
          "number": 125,
        },
      ],
      "_result": true,
    },
    Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$of": Array [
        Object {
          "$course": Object {
            "department": Array [
              "ART",
            ],
            "number": 102,
          },
          "$type": "course",
          "_checked": true,
          "_request": Object {
            "department": Array [
              "ART",
            ],
            "number": 102,
          },
          "_result": true,
          "_taken": true,
        },
        Object {
          "$course": Object {
            "department": Array [
              "ART",
            ],
            "number": 103,
          },
          "$type": "course",
          "_checked": true,
          "_result": false,
        },
      ],
      "$type": "of",
      "_checked": true,
      "_counted": 1,
      "_matches": Array [
        Object {
          "department": Array [
            "ART",
          ],
          "number": 102,
        },
      ],
      "_result": true,
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several of-expressions 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 125,
  },
  Object {
    "department": Array [
      "ART",
    ],
    "number": 102,
  },
]
`;

exports[`computeBoolean can compute the result of several other boolean expressions 1`] = `
Object {
  "$and": Array [
    Object {
      "$booleanType": "or",
      "$or": Array [
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 121,
          },
          "$type": "course",
          "_checked": true,
          "_result": false,
        },
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 125,
          },
          "$type": "course",
          "_checked": true,
          "_request": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 125,
          },
          "_result": true,
          "_taken": true,
        },
      ],
      "$type": "boolean",
      "_checked": true,
      "_matches": Array [
        Object {
          "department": Array [
            "CSCI",
          ],
          "number": 125,
        },
      ],
      "_result": true,
    },
    Object {
      "$booleanType": "or",
      "$or": Array [
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 130,
          },
          "$type": "course",
          "_checked": true,
          "_request": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 130,
          },
          "_result": true,
          "_taken": true,
        },
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 131,
          },
          "$type": "course",
          "_checked": true,
          "_result": false,
        },
      ],
      "$type": "boolean",
      "_checked": true,
      "_matches": Array [
        Object {
          "department": Array [
            "CSCI",
          ],
          "number": 130,
        },
      ],
      "_result": true,
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several other boolean expressions 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 125,
  },
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 130,
  },
]
`;

exports[`computeBoolean can compute the result of several requirement references 1`] = `
Object {
  "$and": Array [
    Object {
      "$requirement": "A",
      "$type": "reference",
      "_checked": true,
      "_matches": Array [
        Object {
          "department": Array [
            "ART",
          ],
          "number": 120,
        },
      ],
      "_result": true,
    },
    Object {
      "$requirement": "C",
      "$type": "reference",
      "_checked": true,
      "_matches": Array [
        Object {
          "department": Array [
            "ART",
          ],
          "number": 104,
        },
        Object {
          "department": Array [
            "ART",
          ],
          "number": 105,
        },
      ],
      "_result": true,
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several requirement references 2`] = `
Array [
  Object {
    "department": Array [
      "ART",
    ],
    "number": 120,
  },
  Object {
    "department": Array [
      "ART",
    ],
    "number": 104,
  },
  Object {
    "department": Array [
      "ART",
    ],
    "number": 105,
  },
]
`;

exports[`computeBoolean can compute the result of several where-expressions 1`] = `
Object {
  "$and": Array [
    Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$distinct": false,
      "$type": "where",
      "$where": Object {
        "$key": "gereqs",
        "$operator": "$eq",
        "$type": "qualification",
        "$value": "WRI",
      },
      "_checked": true,
      "_counted": 1,
      "_matches": Array [
        Object {
          "department": Array [
            "CSCI",
          ],
          "gereqs": Array [
            "WRI",
          ],
          "number": 125,
        },
      ],
      "_result": true,
    },
    Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$distinct": false,
      "$type": "where",
      "$where": Object {
        "$key": "gereqs",
        "$operator": "$eq",
        "$type": "qualification",
        "$value": "BTS-T",
      },
      "_checked": true,
      "_counted": 1,
      "_matches": Array [
        Object {
          "department": Array [
            "ART",
          ],
          "gereqs": Array [
            "BTS-T",
          ],
          "number": 102,
        },
      ],
      "_result": true,
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`computeBoolean can compute the result of several where-expressions 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "gereqs": Array [
      "WRI",
    ],
    "number": 125,
  },
  Object {
    "department": Array [
      "ART",
    ],
    "gereqs": Array [
      "BTS-T",
    ],
    "number": 102,
  },
]
`;

exports[`computeBoolean computes an or-clause even if the first item is false 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
      "_checked": true,
      "_result": false,
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
      "_checked": true,
      "_request": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "_result": true,
      "_taken": true,
    },
  ],
  "$type": "boolean",
}
`;

exports[`computeBoolean computes an or-clause even if the first item is false 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 125,
  },
]
`;

exports[`computeBoolean computes the boolean result of and-clauses 1`] = `
Object {
  "$and": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
      "_checked": true,
      "_request": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "_result": true,
      "_taken": true,
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
      "_checked": true,
      "_request": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "_result": true,
      "_taken": true,
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`computeBoolean computes the boolean result of and-clauses 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 121,
  },
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 125,
  },
]
`;

exports[`computeBoolean computes the boolean result of or-clauses 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
      "_checked": true,
      "_request": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "_result": true,
      "_taken": true,
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
      "_checked": true,
      "_request": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "_result": false,
      "_taken": true,
    },
  ],
  "$type": "boolean",
}
`;

exports[`computeBoolean computes the boolean result of or-clauses 2`] = `
Array [
  Object {
    "department": Array [
      "CSCI",
    ],
    "number": 121,
  },
]
`;
